File: src/whichlang.dart
import 'dart:html';

class CodeBlock {
  String? langName;
  List<String>? langIcon;
  IconType? iconType;
  String? filepath;

  CodeBlock(this.langName, this.langIcon, this.iconType, this.filepath);

  CodeBlock.fromClasses(Iterable<String> classes) {
    var forceCurrentLang = false;

    for (final cls in classes) {
      if (!forceCurrentLang) langName ??= langByHljsClass(cls);

      final match = keyValuePattern.firstMatch(cls);
      if (match != null) {
        final key = match.namedGroup('key')!;
        final value = match.namedGroup('value')!;
        switch (key) {
          case 'icon':
            var iconName;

            iconName = value.replaceFirst(RegExp(r'^\.'), '');
            if (iconName != value) {
              iconType = IconType.classNames;
              langIcon = iconName.split('.');
              break;
            }

            iconName = value.replaceFirst(RegExp(r'^@'), '');
            if (iconName != value) {
              iconType = IconType.url;
              langIcon = [iconName];
              break;
            }

          case 'lang':
            langName = value;

          case 'filepath':
            filepath = value;
        }
      } else {
        switch (cls) {
          case 'nolang':
            langName = null;
            forceCurrentLang = true;
        }
      }
    }
  }

  bool anySet() {
    return [langName, langIcon, iconType, filepath].any((e) => e != null);
  }

  String toString() {
    return 'CodeBlock(langName: $langName, langIcon: $langIcon, iconType: $iconType, filepath: $filepath)';
  }
}

enum IconType { url, classNames }

final hljsLangPattern = RegExp(r'^language-');
final keyValuePattern = RegExp(r'(?<key>.*?)=(?<value>.*)');

/// Transforms language names which are not manually set with `lang=`.
String transformLangName(String name) {
  return name.toUpperCase();
}

/// `highlightjs` assigns (multiple) classes of the form `language-x` to code blocks.
/// With that we can find out which language or family member of that language has been
/// specified in the Markdown code block.
String? langByHljsClass(String className) {
  final String name = className.replaceFirst(hljsLangPattern, '');
  if (className == name) {
    return null;
  }
  return transformLangName(name);
}

/// Inserts a `div` with class `wl-codeblockheader` in front of each `pre` element which should
/// be the parent element of `codeBlockElement`. When a language or an icon is defined, both of
/// these are wrapped within a `div` with the class `wl-langinfo`. Next to that element there may
/// be a `i` or `svg` which represents the icon.
void modifyCodeBlock(Element codeBlockElement, CodeBlock codeBlockData) {
  if (!codeBlockData.anySet()) return;

  var codeBlockHeader = document.createElement('div');
  codeBlockHeader.classes.add('wl-codeblockheader');

  var langInfo = document.createElement('div');
  langInfo.classes.add('wl-langinfo');

  final pre = codeBlockElement.parent;
  if (pre == null) {
    return;
  }
  pre.before(codeBlockHeader);

  // icon
  if (codeBlockData.langIcon != null) {
    var langIcon = null;
    switch (codeBlockData.iconType!) {
      case IconType.classNames:
        langIcon = document.createElement('i');
        langIcon.classes.add('wl-icon');
        codeBlockData.langIcon!.forEach((e) => langIcon.classes.add(e));
        langInfo.append(langIcon);

      case IconType.url:
        var image = document.createElement('img');
        image.setAttribute('src', codeBlockData.langIcon!.first);
        langInfo.append(image);
    }
  }

  // language name
  if (codeBlockData.langName != null) {
    var langName = document.createElement('span');
    langName.classes.add('wl-lang-name');
    langName.appendText(codeBlockData.langName!);
    langInfo.append(langName);
  }

  if (langInfo.hasChildNodes()) codeBlockHeader.append(langInfo);

  // filepath
  if (codeBlockData.filepath != null) {
    var filepath = document.createElement('span');
    filepath.classes.add('wl-filepath');
    filepath.appendText(codeBlockData.filepath!);
    codeBlockHeader.append(filepath);
  }
}

void main() {
  final codeBlocks = Map<Element, CodeBlock>.fromIterable(
      document.querySelectorAll('pre > code').map(
          (codeBlock) => (codeBlock, CodeBlock.fromClasses(codeBlock.classes))),
      key: (item) => item.$1,
      value: (item) => item.$2);

  codeBlocks.entries.forEach((e) => modifyCodeBlock(e.key, e.value));
}

File: src/whichlang.ts
import { config } from "./config.ts";

type Banner = {
  icon?: HTMLElement;
  language?: string;
  filePath?: string;
};

function codeBlockBanner(bannerData: Banner): HTMLDivElement {
  const banner = document.createElement("div");
  banner.classList.add("wl-banner");

  const icon = document.createElement("span");
  icon.classList.add("wl-icon");

  if (bannerData.icon !== undefined) {
    icon.append(bannerData.icon.cloneNode(true));
  }
  banner.append(icon);

  const filePath = document.createElement("span");
  filePath.classList.add("wl-filepath");

  if (bannerData.filePath !== undefined) {
    filePath.textContent = bannerData.filePath;
  }
  banner.append(filePath);

  return banner;
}

function transformHljsIdentifier(hljsIdentifier: string): string {
  return hljsIdentifier.toUpperCase();
}

function languageNameByHljsClass(className: string): string | undefined {
  const hljsIdentifier = className.replace(/^language-/, "");
  if (className == hljsIdentifier) return undefined;
  return transformHljsIdentifier(hljsIdentifier);
}

function parseAttribute(input: string): [string, string] | undefined {
  return input.split("=", 2) as [string, string];
}

for (const codeBlock of document.querySelectorAll("pre > code")) {
  let language: string | undefined;
  let filePath: string | undefined;
  let icon: HTMLElement | undefined;

  let forceCurrentIcon: boolean = false;

  let createBanner = !config.explicitBanner;

  const classes = codeBlock.classList;
  for (const cls of classes) {
    language ??= languageNameByHljsClass(cls);

    const pair = parseAttribute(cls);
    if (pair === undefined) continue;
    const [key, value] = pair;
    switch (key) {
      case "icon":
        if (value.startsWith("%")) {
          icon = config.icons[value.slice(1)];
        } else if (value.startsWith(".")) {
          const iconClasses = value.split(".").filter((x) => x.trim() !== "");
          icon = document.createElement("i");
          icon.classList.add(...iconClasses);
        } else if (value.startsWith("@")) {
          const url = value.slice(1);
          icon = document.createElement("img");
          icon.setAttribute("src", url);
          icon.setAttribute("width", "20");
          icon.setAttribute("height", "20");
        } else {
          icon = undefined;
        }
        forceCurrentIcon = true;
        break;
      case "fp":
      case "filepath":
        filePath = value;
        break;
      case "banner":
        switch (value) {
          case "yes":
            createBanner = true;
            break;
          case "no":
            createBanner = false;
            break;
          default:
            console.warn(
              `Invalid value for banner attribute of code block: ${value}`,
            );
        }
        break;
      default:
        console.debug(`Ignoring pair ${cls}`);
    }
  }

  if (!createBanner) continue;

  if (icon === undefined && !forceCurrentIcon && language !== undefined) {
    icon = config.icons[language.toLowerCase()];
  }

  if (icon !== undefined || filePath !== undefined) {
    const banner: Banner = { icon, language, filePath };
    const element = codeBlockBanner(banner);
    codeBlock.parentElement?.before(element);
  }
}
